azuread/resource_user.go

```diff
@@ -0,0 +1,180 @@
+package azuread
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azuread/azuread/helpers/ar"
+	"github.com/terraform-providers/terraform-provider-azuread/azuread/helpers/p"
+	"github.com/terraform-providers/terraform-provider-azuread/azuread/helpers/validate"
+)
+
+func resourceUser() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceUserCreate,
+		Read:   resourceUserRead,
+		Update: resourceUserUpdate,
+		Delete: resourceUserDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"user_principal_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.StringIsEmailAddress,
+			},
+
+			"display_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validate.NoEmptyStrings,
+			},
+
+			"mail_nickname": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validate.NoEmptyStrings,
+			},
+
+			"account_enabled": {
+				Type:     schema.TypeBool,
+				Required: true,
+			},
+
+			"password": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Sensitive:    true,
+				ValidateFunc: validation.StringLenBetween(1, 16), //currently the max length for AAD passwords is 16
+			},
+
+			"force_password_change": {
+				Type:     schema.TypeBool,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceUserCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).usersClient
+	ctx := meta.(*ArmClient).StopContext
+
+	userPrincipalName := d.Get("user_principal_name").(string)
+	displayName := d.Get("display_name").(string)
+	mailNickName := d.Get("mail_nickname").(string)
+	accountEnabled := d.Get("account_enabled").(bool)
+	password := d.Get("password").(string)
+	forcePasswordChange := d.Get("force_password_change").(bool)
+
+	userCreateParameters := graphrbac.UserCreateParameters{
+		AccountEnabled: &accountEnabled,
+		DisplayName:    &displayName,
+		MailNickname:   &mailNickName,
+		PasswordProfile: &graphrbac.PasswordProfile{
+			ForceChangePasswordNextLogin: &forcePasswordChange,
+			Password:                     &password,
+		},
+		UserPrincipalName: &userPrincipalName,
+	}
+
+	user, err := client.Create(ctx, userCreateParameters)
+	if err != nil {
+		return fmt.Errorf("Error creating User %q: %+v", userPrincipalName, err)
+	}
+
+	objectId := user.ObjectID
+
+	resp, err := client.Get(ctx, *objectId)
+	if err != nil {
+		return fmt.Errorf("Error retrieving User (%q) with ObjectID %q: %+v", userPrincipalName, *objectId, err)
+	}
+
+	d.SetId(*resp.ObjectID)
```

Comment:
As above, we should nil check this dereference

Reply:
Updated as requested.